feat(minifier): implement folding simple arrow fns

camc314 opened the pull request:
basically
```ts
const foo = () => {
    return baz
}
```
becomes
```ts
const foo = () => baz 
```

camc314 commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6875?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6875** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6875?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6871** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6871?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6870** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6870?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6869** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6869?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @camc314 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented:
Could this introduce an ASI hazard?

camc314 commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
not sure whether this is the right file for this minification - i don't think closure implements this

Boshen replied:
It does, you need to find the right place for this code https://closure-compiler.appspot.com/home#code%3Dconst%2520foo%2520%253D%2520()%2520%253D%253E%2520%257B%250A%2520%2520%2520%2520return%2520baz%250A%257D

camc314 replied:
look like it happens in here >>

https://github.com/google/closure-compiler/blob/4c87a6855aa66d6b1bddfee76aba97d45f9ae065/src/com/google/javascript/jscomp/SubstituteEs6Syntax.java#L66-L78

should i create a new minifier pass for this optimization (substitute_es6_syntax) ?

camc314 replied:
the other optimization here (i think)
is https://github.com/google/closure-compiler/blob/4c87a6855aa66d6b1bddfee76aba97d45f9ae065/src/com/google/javascript/jscomp/SubstituteEs6Syntax.java#L51
which would happen (for us) in codegen if i understand correctly?

Boshen replied:
😅 that's a really weird place. I think you've put the code in the right place.

camc314 commented:
> Could this introduce an ASI hazard? We may want to insert semicolons to be safe.

good question - i think this would be handled in oxc_codegen?

Boshen commented on crates/oxc_minifier/src/ast_passes/peephole_substitute_alternate_syntax.rs:
```diff
@@ -121,6 +121,26 @@ impl<'a> Traverse<'a> for PeepholeSubstituteAlternateSyntax {
+                {
+                    if let Some(body) = arrow_expr.body.statements.first_mut() {
+                        let result = if let Statement::ReturnStatement(ret_stmt) = body {
+                            ret_stmt.argument.as_mut().map(|arg| ctx.ast.move_expression(arg))
```
This may move the code without setting changed to true?

camc314 replied:
don't think so but i improved this slightly to make it more readable/clear

Boshen commented:
### Merge activity

* **Oct 25, 6:25 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camc314 committed:
feat(minifier): implement folding simple arrow fns (#6875)

basically
```ts
const foo = () => {
    return baz
}
```
becomes
```ts
const foo = () => baz
```
